test/hotspot/jtreg/runtime/Monitor/MonitorWithDeadObjectTest.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (c) 2022, 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+/*
+ * @summary This test checks that ObjectMonitors with dead objects don't
+            cause asserts, crashes, or failures when various sub-systems
+            in the JVM find them.
+ * @requires os.family != "windows"
+ * @library /testlibrary /test/lib
+ * @modules jdk.management
+ */
+
+/*
+ * @test id=DetachThread
+ * @run main/othervm/native MonitorWithDeadObjectTest 0
+ */
+
+/*
+ * @test id=DumpThreadsBeforeDetach
+ * @run main/othervm/native MonitorWithDeadObjectTest 1
+ */
+
+/*
+ * @test id=DumpThreadsAfterDetachBeforeJoin
+ * @run main/othervm/native MonitorWithDeadObjectTest 2
+ */
+
+/*
+ * @test id=DumpThreadsAfterDetachAfterJoin
+ * @run main/othervm/native MonitorWithDeadObjectTest 3
+ */
+
+import java.lang.management.ManagementFactory;
+import java.lang.management.ThreadMXBean;
+
+public class MonitorWithDeadObjectTest {
+    public static native void createMonitorWithDeadObject();
+    public static native void createMonitorWithDeadObjectNoJoin();
+    public static native void createMonitorWithDeadObjectDumpThreadsBeforeDetach();
+
+    public static native void joinTestThread();
+
+    static {
+        System.loadLibrary("MonitorWithDeadObjectTest");
+    }
+
+    private static void dumpThreadsWithLockedMonitors() {
+        ThreadMXBean threadBean = ManagementFactory.getThreadMXBean();
+        threadBean.dumpAllThreads(true, false);
+    }
+
+    private static void testDetachThread() {
+        // Create an ObjectMonitor with a dead object from an
+        // attached thread.
```

Comment:
And why add these wrapper methods that simply call one other method?

Reply:
> Unclear what the "Detach" in the method name has to do with anything. ??

This test case provokes the assert we hit when the monitor is visited inside DetachCurrentThread. I updated the comment to state that.

> And why add these wrapper methods that simply call one other method?

Because I find this structure more cohesive and better structured. I have four functions representing the four tests. The fact that two of them in turn only call one function is an implementation detail. I don't want to push the call to `createMonitorWithDeadObject` down into the main function, because then I also have to move the comment there, and suddenly the main function becomes more then just a super simple dispatch function.